sklearn/utils/_param_validation.py

```diff
@@ -499,6 +499,30 @@ def __str__(self):
         )
 
 
+class HasMethod(_Constraint):
+    """Constraint representing objects that expose a specific method.
+
+    It is useful for parameters following a protocol and where we don't want to impose
+    an affiliation to a specific module or class.
+
+    Parameters
+    ----------
+    method : str
+        The method that the object is expected to expose.
+    """
+
+    @validate_params({"method": [str]})
+    def __init__(self, method):
```

Comment:
In the future, I think we will need to extend this for two methods. For example,  `MultiOutputRegressor` requires two methods: `fit` and `predict`:

https://github.com/scikit-learn/scikit-learn/blob/3c75d36c6585b175f29f207c20cced1222b4d1b7/sklearn/multioutput.py#L255-L256

Reply:
Would it be worth treating `HasMethod` similar to `StrOptions` in that you can provide a set of methods to check against (and calling it `HasMethods`). Then `Is{Classifier,Regressor,Transformer,Estimator}` could be special cases of that.

Reply:
> Would it be worth treating HasMethod similar to StrOptions

I think so yes, with a slight difference from ``StrOptions``, since the object would need to implement all listed methods and not one of the listed methods.

> Basically, I am wondering if we should not create a protocol for the minimal classifier, regressor and transformer.

It would be interesting to explore, but I guess we can't start with what's in this PR first explore that later.